File: backend/cache.py
import sqlite3
import json
import os
from datetime import datetime, timedelta
from typing import Optional, Dict, Any, List

DB_PATH = os.path.join(os.path.dirname(__file__), "github_cache.db")
CACHE_EXPIRATION_HOURS = 24

def get_db_connection():
    conn = sqlite3.connect(DB_PATH)
    conn.row_factory = sqlite3.Row
    return conn

def init_db():
    """Initializes the SQLite database tables if they do not exist."""
    conn = get_db_connection()
    cursor = conn.cursor()
    
    # Profile Cache Table
    cursor.execute("""
        CREATE TABLE IF NOT EXISTS profiles (
            username TEXT PRIMARY KEY,
            data_json TEXT NOT NULL,
            cached_at TEXT NOT NULL
        )
    """)
    
    # Repositories Cache Table
    cursor.execute("""
        CREATE TABLE IF NOT EXISTS repositories (
            username TEXT,
            repo_name TEXT,
            data_json TEXT NOT NULL,
            cached_at TEXT NOT NULL,
            PRIMARY KEY (username, repo_name)
        )
    """)
    
    conn.commit()
    conn.close()

def is_expired(cached_time_str: str) -> bool:
    """Checks if the cached data has expired."""
    try:
        cached_time = datetime.fromisoformat(cached_time_str)
        return datetime.now() - cached_time > timedelta(hours=CACHE_EXPIRATION_HOURS)
    except Exception:
        return True

def get_cached_profile(username: str) -> Optional[Dict[str, Any]]:
    """Retrieves the cached profile if it exists and is not expired."""
    username = username.lower().strip()
    conn = get_db_connection()
    cursor = conn.cursor()
    cursor.execute("SELECT data_json, cached_at FROM profiles WHERE username = ?", (username,))
    row = cursor.fetchone()
    conn.close()
    
    if row and not is_expired(row["cached_at"]):
        return json.loads(row["data_json"])
    return None

def set_cached_profile(username: str, profile_data: Dict[str, Any]):
    """Caches profile data."""
    username = username.lower().strip()
    conn = get_db_connection()
    cursor = conn.cursor()
    now_str = datetime.now().isoformat()
    cursor.execute(
        "INSERT OR REPLACE INTO profiles (username, data_json, cached_at) VALUES (?, ?, ?)",
        (username, json.dumps(profile_data), now_str)
    )
    conn.commit()
    conn.close()

def get_cached_repos(username: str) -> Optional[List[Dict[str, Any]]]:
    """Retrieves cached repository data for a user if it exists and is not expired."""
    username = username.lower().strip()
    conn = get_db_connection()
    cursor = conn.cursor()
    
    # Check if there is any repo and if it's expired by checking the first match
    cursor.execute("SELECT data_json, cached_at FROM repositories WHERE username = ? LIMIT 1", (username,))
    first_row = cursor.fetchone()
    
    if not first_row or is_expired(first_row["cached_at"]):
        conn.close()
        return None
        
    cursor.execute("SELECT data_json FROM repositories WHERE username = ?", (username,))
    rows = cursor.fetchall()
    conn.close()
    
    return [json.loads(row["data_json"]) for row in rows]

def set_cached_repos(username: str, repos_data: List[Dict[str, Any]]):
    """Caches a list of repositories for a user."""
    username = username.lower().strip()
    conn = get_db_connection()
    cursor = conn.cursor()
    now_str = datetime.now().isoformat()
    
    # Clear old entries first
    cursor.execute("DELETE FROM repositories WHERE username = ?", (username,))
    
    for repo in repos_data:
        repo_name = repo.get("name", "")
        cursor.execute(
            "INSERT OR REPLACE INTO repositories (username, repo_name, data_json, cached_at) VALUES (?, ?, ?, ?)",
            (username, repo_name, json.dumps(repo), now_str)
        )
    conn.commit()
    conn.close()

def clear_cache(username: str):
    """Clears cache for a given user."""
    username = username.lower().strip()
    conn = get_db_connection()
    cursor = conn.cursor()
    cursor.execute("DELETE FROM profiles WHERE username = ?", (username,))
    cursor.execute("DELETE FROM repositories WHERE username = ?", (username,))
    conn.commit()
    conn.close()

# Initialize on import
init_db()

File: backend/cache.js
const sqlite3 = require('sqlite3').verbose();
const path = require('path');

const dbPath = path.join(__dirname, 'github_cache.db');
const db = new sqlite3.Database(dbPath);

// Promisify Database functions
const dbRun = (query, params = []) => {
    return new Promise((resolve, reject) => {
        db.run(query, params, function (err) {
            if (err) reject(err);
            else resolve(this);
        });
    });
};

const dbGet = (query, params = []) => {
    return new Promise((resolve, reject) => {
        db.get(query, params, (err, row) => {
            if (err) reject(err);
            else resolve(row);
        });
    });
};

const dbAll = (query, params = []) => {
    return new Promise((resolve, reject) => {
        db.all(query, params, (err, rows) => {
            if (err) reject(err);
            else resolve(rows);
        });
    });
};

const initDb = async () => {
    try {
        await dbRun(`
            CREATE TABLE IF NOT EXISTS users (
                username TEXT PRIMARY KEY,
                profile_data TEXT,
                last_updated TIMESTAMP DEFAULT CURRENT_TIMESTAMP
            )
        `);
        await dbRun(`
            CREATE TABLE IF NOT EXISTS repos (
                username TEXT,
                repo_name TEXT,
                repo_data TEXT,
                PRIMARY KEY (username, repo_name)
            )
        `);
    } catch (e) {
        console.error("Database table creation failed", e);
    }
};

// Initialize DB on module load
initDb().catch(err => console.error("Database initialization failed", err));

const getCachedProfile = async (username) => {
    try {
        const row = await dbGet("SELECT profile_data, last_updated FROM users WHERE username = ?", [username.toLowerCase()]);
        if (!row) return null;
        
        // Expire cache after 24 hours
        const diffHours = (new Date() - new Date(row.last_updated)) / 36e5;
        if (diffHours > 24) return null;
        
        return JSON.parse(row.profile_data);
    } catch (e) {
        console.error("Cache getCachedProfile error", e);
        return null;
    }
};

const getCachedRepos = async (username) => {
    try {
        const rows = await dbAll("SELECT repo_data FROM repos WHERE username = ?", [username.toLowerCase()]);
        if (!rows || rows.length === 0) return null;
        return rows.map(r => JSON.parse(r.repo_data));
    } catch (e) {
        console.error("Cache getCachedRepos error", e);
        return null;
    }
};

const saveProfileCache = async (username, profileData) => {
    try {
        await dbRun(
            "INSERT OR REPLACE INTO users (username, profile_data, last_updated) VALUES (?, ?, CURRENT_TIMESTAMP)",
            [username.toLowerCase(), JSON.stringify(profileData)]
        );
    } catch (e) {
        console.error("Cache saveProfileCache error", e);
    }
};

const saveReposCache = async (username, reposData) => {
    try {
        // Clear existing repos cache for this user first
        await dbRun("DELETE FROM repos WHERE username = ?", [username.toLowerCase()]);
        for (const repo of reposData) {
            await dbRun(
                "INSERT INTO repos (username, repo_name, repo_data) VALUES (?, ?, ?)",
                [username.toLowerCase(), repo.name, JSON.stringify(repo)]
            );
        }
    } catch (e) {
        console.error("Cache saveReposCache error", e);
    }
};

module.exports = {
    getCachedProfile,
    getCachedRepos,
    saveProfileCache,
    saveReposCache
};
